analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs

```diff
@@ -25,14 +25,27 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn
     public sealed record ProgramState : IEquatable<ProgramState>
     {
         public static readonly ProgramState Empty = new();
-        private ProgramState[] toArray;
+        private readonly ImmutableDictionary<IOperation, SymbolicValue> operationValue;
+        private readonly ImmutableDictionary<ISymbol, SymbolicValue> symbolValue;
+        private readonly ImmutableDictionary<CaptureId, IOperation> captureOperation;
+        private readonly ImmutableHashSet<ISymbol> preservedSymbols;
+        private readonly ImmutableStack<ExceptionState> exceptions;
 
-        private ImmutableDictionary<IOperation, SymbolicValue> OperationValue { get; init; }     // Current SymbolicValue result of a given operation
-        private ImmutableDictionary<ISymbol, SymbolicValue> SymbolValue { get; init; }
+        private ProgramState[] toArray;
+        private int? operationValueHashCode;
+        private int? symbolValueHashCode;
+        private int? captureOperationHashCode;
+        private int? preservedSymbolsHashCode;
+        private int? exceptionsHashCode;
```

Comment:
Can we make a private generic class that would wrap this pair of value & its hashCode to duplicate it? And move as much logic inside, like the one that is inside `init`.

And maybe delegate also the hashCode computation there.

Reply:
In theory, yes. But we need to make the resulting struct a mutable field in the record. If we would make the field readonly, the code would compile but then we end up with defensive copy problems because the struct can not be readonly (hashcode needs to be changed on the first call to GetHashCode). The problem is described here: https://levelup.gitconnected.com/defensive-copy-in-net-c-38ae28b828

I also don't see how we could implement the `??=` logic easily.

So we could do it, but we have to leave some comments to make sure we don't break anything.